src/resolvers/libsodium.rs

```diff
@@ -0,0 +1,280 @@
+extern crate sodiumoxide;
+
+use byteorder::{ByteOrder, LittleEndian};
+
+use params::{CipherChoice, DHChoice, HashChoice};
+use types::{Cipher, Dh, Hash, Random};
+use CryptoResolver;
+
+use self::sodiumoxide::crypto::aead::chacha20poly1305_ietf as sodium_chacha20poly1305;
+use self::sodiumoxide::crypto::hash::sha256 as sodium_sha256;
+use self::sodiumoxide::crypto::scalarmult::curve25519 as sodium_curve25519;
+
+#[derive(Default)]
+pub struct SodiumResolver;
+
+impl CryptoResolver for SodiumResolver {
+    fn resolve_rng(&self) -> Option<Box<Random + Send>> {
+        None
+    }
+
+    fn resolve_dh(&self, choice: &DHChoice) -> Option<Box<Dh + Send>> {
+        match *choice {
+            DHChoice::Curve25519 => Some(Box::new(SodiumDh25519::default())),
+            _ => None,
+        }
+    }
+
+    fn resolve_hash(&self, choice: &HashChoice) -> Option<Box<Hash + Send>> {
+        match *choice {
+            HashChoice::SHA256 => Some(Box::new(SodiumSha256::default())),
+            _ => None,
+        }
+    }
+
+    fn resolve_cipher(&self, choice: &CipherChoice) -> Option<Box<Cipher + Send>> {
+        match *choice {
+            CipherChoice::ChaChaPoly => Some(Box::new(SodiumChaChaPoly::default())),
+            _ => None,
+        }
+    }
+}
+
+// Elliptic curve 25519.
+pub struct SodiumDh25519 {
+    privkey: sodium_curve25519::Scalar,
+    pubkey: sodium_curve25519::GroupElement,
+}
+
+impl SodiumDh25519 {
+    fn convert_to_private_key(key: &mut [u8; 32]) {
+        key[0] &= 248;
```

Comment:
this function is only used once, can just inline it in `generate()`?

Reply:
IMHO conversion logic should be encapsulated and not mixed up with key generation.
Since it's used only once, it'll be inlined by the compiler, and `convert_*` call describes what's going on better than some binary operations, even with the comment.

`libsodium` contains separate function for such a conversion, and as soon as it's added to the `sodiumoxide` (I'm going to add corresponding PR), this method will be removed.

Reply:
Fair argument, will be a moot point (and aligned with other x25519 libraries like dalek) once they add the function in themselves :).